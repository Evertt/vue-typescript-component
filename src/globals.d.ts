/**
 * global type declarations in this project
 * should not expose to userland
 */

declare const process: {
	env: {
		NODE_ENV: string,
	},
}

declare interface Array<T> {
	mapToObject(transform: (obj: any) => any): any
}
