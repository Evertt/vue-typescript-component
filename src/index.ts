export * from './prop'
export * from './watch'
export * from './lifecycle'
export * from './component'
