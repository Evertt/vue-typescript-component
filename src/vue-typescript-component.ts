import Vue = require('vue')

export class WatchOptions {
	expression: string
	options: { handler: Vue.WatchHandler<Vue> } & Vue.WatchOptions
}

export class VueTypescriptComponentData {
	props: {[index: string]: Vue.PropOptions} = {}
	watch: {[index: string]: WatchOptions[]} = {}
	injected: {[index: string]: Object} = {}
}

// implementation Object.assign (targets only the scope of this project)
function objAssign<T, U, V>(target: T, source1: U, source2?: V): T & U & V {
	if (source1) {
		for (let n of Object.getOwnPropertyNames(source1)) {
			Object.defineProperty(target, n, Object.getOwnPropertyDescriptor(source1, n))
		}
	}
	if (source2) {
		for (let n of Object.getOwnPropertyNames(source2)) {
			Object.defineProperty(target, n, Object.getOwnPropertyDescriptor(source2, n))
		}
	}
	return <T & U & V>target
}

function getOrCreateComponent(target: any): VueTypescriptComponentData {
	if (typeof target !== 'object') {
		throw 'VueTypescriptComponent decorator is only allowed for non-static members'
	}
	if (!target.hasOwnProperty('$$vueTypescriptComponentData')) {
		let superInjected = {}
		if (target.$$vueTypescriptComponentData) {
			superInjected = objAssign({}, target.$$vueTypescriptComponentData.injected)
		}
		target.$$vueTypescriptComponentData = new VueTypescriptComponentData()
		target.$$vueTypescriptComponentData.injected = superInjected
	}
	return target.$$vueTypescriptComponentData
}

/** Use field as property
 * If field is initialized, sets default value and type. Otherwise mark as required
 */
export function prop(options: Vue.PropOptions = {}) {
	return (target: any, member: string) => {
		getOrCreateComponent(target).props[member] = options
	}
}

/** Ignore field
 * Use on injected fields to make sure they do not appear in data()
 */
export function injected() {
	return (target: any, member: string) => {
		getOrCreateComponent(target).injected[member] = {}
	}
}

/** Use method(val, oldVal) to watch expression
 * The method is still part of the `methods` object and it is possible to use multiple @watch
 * annotations to watch multiple expressions with one function
 */
export function watch(expression: string, options: Vue.WatchOptions = {}) {
	return (target: any, member: string) => {
		const wo = new WatchOptions()
		wo.expression = expression
		wo.options = <{ handler: Vue.WatchHandler<Vue> } & WatchOptions>options
		const watch = getOrCreateComponent(target).watch
		if (!watch[member]) {
			watch[member] = []
		}
		watch[member].push(wo)
	}
}

export interface NoArgumentConstructable {
	new (): any
	name?: any
	vueComponentOptions?: Vue.ComponentOptions<Vue>
}

const lifecycleHooks = ['beforeCreate', 'created', 'beforeMount', 'mounted',
		'beforeUpdate', 'updated', 'activated', 'deactivated',
		'beforeDestroy', 'destroyed',
		'render', /* not a lifecyle hook, but handled identically */
]

/** Create property constructor.vueComponentOptions based on method/field annotations
 *  If provided, use options as the base value (.data is always overridden)
 */
export function component(options: Vue.ComponentOptions<Vue> = {}) {
	return (cls: NoArgumentConstructable) => {
		const d = getOrCreateComponent(cls.prototype)
		const superOptions: Vue.ComponentOptions<Vue> = <Vue.ComponentOptions<Vue>>cls.vueComponentOptions || {}
		const obj = new cls()
		cls.vueComponentOptions = options
		options.name = options.name || (<any>cls).name
		options.methods = objAssign({}, superOptions.methods, options.methods)
		options.computed = objAssign({}, superOptions.computed, options.computed)
		options.watch = objAssign({}, superOptions.watch, options.watch)
		options.props = objAssign({}, superOptions.props, options.props)
		// to get rid of Index signature of object type implicitly has an 'any' type.
		const props = <{ [key: string]: Vue.PropOptions }>options.props
		options.data = () => {
			const newData = new cls()
			const r: any = {}
			for (let n of Object.getOwnPropertyNames(newData)) {
				if (!props[n] && n[0] !== '_' && n[0] !== '$' && !d.injected[n]) {
					r[n] = newData[n]
				}
			}
			return r
		}

		for (let n of Object.getOwnPropertyNames(d.props)) {
			props[n] = d.props[n]
		}
		// create data; add default values and types to props with initialisers
		for (let n of Object.getOwnPropertyNames(obj)) {
			if (props[n]) {
				if (typeof obj[n] === 'object') {
					props[n].default = () => { return new cls()[n] }
				} else {
					props[n].default = obj[n]
				}
				if (props[n].type === undefined) {
					props[n].type = Object.getPrototypeOf(obj[n]).constructor
				}
			}
		}
		// set props without default value to required
		for (let n in props) {
			if (props[n].required === undefined && props[n].default === undefined) {
				props[n].required = true
			}
		}
		// methods
		for (let n of Object.getOwnPropertyNames(cls.prototype)) {
			if (n === 'constructor' || n === '$$vueTypescriptComponentData') {
				continue
			}
			const pd = Object.getOwnPropertyDescriptor(cls.prototype, n)
			if (pd.set) {
				options.computed![n] = { set: pd.set, get: pd.get }
			} else if (pd.get) {
				options.computed![n] = pd.get
			} else if (lifecycleHooks.indexOf(n) !== -1) {
				(<any>options)[n] = pd.value
			} else {
				options.methods![n] = pd.value
			}
		}
		// watch
		for (let n of Object.getOwnPropertyNames(d.watch)) {
			if (!options.methods![n]) {
				throw '@watch must decorate a method: ' + cls.name + '.' + n
			}
			for (let o of d.watch[n]) {
				if (options.watch![o.expression] && superOptions.watch![o.expression] !== options.watch![o.expression]) {
					throw 'duplicate watch expression: ' + cls.name + '.' + o.expression
				}
				o.options.handler = obj[n]
				options.watch![o.expression] = o.options
			}
		}
	}
}
