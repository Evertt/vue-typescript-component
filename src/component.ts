import Vue = require('vue')
import { VueClass } from './vue-class'
import { objAssign } from './util'
import { decorators } from './decorators'

export type ComponentOptions = Vue.ComponentOptions<Vue>

Array.prototype.mapToObject =
	function(transform: (obj: any) => any): any {
		let transformed: any = {}
		for (let elem of this) {
			objAssign(transformed, transform(elem))
		}
		return transformed
	}

export interface NoArgumentConstructable {
	new (): any
	name?: any
	vueComponentOptions?: ComponentOptions
}

function isNoArgumentConstructable(arg: any): arg is NoArgumentConstructable {
	return arg instanceof Function
}

let makeObject = (obj: any) => (key: string) => ({ [key]: obj[key] })
let getMethod  = (obj: any) => (key: string) => Object.getOwnPropertyDescriptor(obj, key).value

function makeComponentDecorator(options: ComponentOptions = {}): ComponentDecorator {
	return (Component: NoArgumentConstructable) => {
		const component = new Component()
		const prototype = Component.prototype

		const superProto = Object.getPrototypeOf(prototype)
		const Super = superProto instanceof Vue
			? superProto.constructor as VueClass
			: Vue

		let andDecorate = (obj: any) => (key: string) =>
			key == 'constructor' || ! (decorators[key] || []).reduce(
				(result, decorator) =>
					decorator.target == obj || decorator.target == obj.__proto__
						? decorator.fn(options, obj) || result : result, false,
			)

		let componentKeys = Object.getOwnPropertyNames(component).filter(andDecorate(component))
		let prototypeKeys = Object.getOwnPropertyNames(prototype).filter(andDecorate(prototype))

		let decKeys = Object.keys(decorators)
		let allKeys = componentKeys.concat(prototypeKeys)
		decKeys.filter(key => allKeys.indexOf(key) === -1).filter(andDecorate(component))

		let staticKeys = Object.getOwnPropertyNames(Component)
		let superStaticKeys = Object.getOwnPropertyNames(Super)
		let componentStaticKeys = staticKeys.filter( key => superStaticKeys.indexOf(key) === -1)

		componentKeys = componentKeys.filter(andDecorate(component))
		prototypeKeys = prototypeKeys.filter(andDecorate(prototype))

		getData(componentKeys, options, component)
		getMethods(prototypeKeys, options, prototype)
		getComputed(prototypeKeys, options, prototype)
		getOptions(componentStaticKeys, options, Component)

		const superOptions = Component.vueComponentOptions || {}
		options.name       = options.name || (<any>Component).name
		options.methods    = objAssign( {}, superOptions.methods,  options.methods  )
		options.computed   = objAssign( {}, superOptions.computed, options.computed )
		options.watch      = objAssign( {}, superOptions.watch,    options.watch    )
		options.props      = objAssign( {}, superOptions.props,    options.props    )

		Component.vueComponentOptions = options

		if (process.env.NODE_ENV !== 'test') return Super.extend(options)
	}
}

function getData(keys: string[], options: ComponentOptions, component: any) {
	let dataKeys = keys.filter(key => key[0] !== '_' && key[0] !== '$')
	options.data = () => dataKeys.mapToObject(makeObject(component))
}

function getMethods(keys: string[], options: ComponentOptions, prototype: any) {
	let dataKeys = keys.filter(key => key !== 'constructor')

	let methods = (options.methods || (options.methods = {}))

	for (let key of dataKeys) {
		let method = getMethod(prototype)(key)
		if (method) { methods[key] = method }
	}
}

function getComputed(keys: string[], options: ComponentOptions, prototype: any) {
	let computedKeys = keys.filter(key => Object.getOwnPropertyDescriptor(prototype, key).get)

	options.computed = computedKeys.mapToObject( key => {
		let pd = Object.getOwnPropertyDescriptor(prototype, key)
		return { [key]: { get: pd.get, set: pd.set } }
	})
}

function getOptions(keys: string[], options: ComponentOptions, Component: any) {
	objAssign(options, keys.mapToObject(makeObject(Component)))
}

/** Create property constructor.vueComponentOptions based on method/field annotations
 *  If provided, use options as the base value (.data is always overridden)
 */
export type ComponentDecorator = (cls: NoArgumentConstructable) => void
export function Component(cls: NoArgumentConstructable): void
export function Component(options: ComponentOptions): ComponentDecorator
export function Component(options: ComponentOptions | NoArgumentConstructable): ComponentDecorator | void {
	if (isNoArgumentConstructable(options)) {
		return makeComponentDecorator()(options)
	} else {
		return makeComponentDecorator(options)
	}
}

export const component = Component
