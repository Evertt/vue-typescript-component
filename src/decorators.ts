import { ComponentOptions } from './component'

export type Decorator = (options: ComponentOptions) => void | boolean

export let decorators: { [key: string]: Decorator[] } = {}

export function addDecorator(key: string, decorator: Decorator) {
	let list = decorators[key] || []
	list.push(decorator)
	decorators[key] = list
}
