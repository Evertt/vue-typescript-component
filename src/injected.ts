import { addDecorator } from './decorators'

export function injected(target: any, key: string): void {
    addDecorator(key, () => true)
}

export const Injected = injected
