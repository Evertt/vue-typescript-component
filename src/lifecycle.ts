import { warn } from './util'
import { addDecorator } from './decorators'

const lifeCycleHooks = [
		'beforeCreate', 'created',
		'beforeMount', 'mounted',
		'activated', 'deactivated',
		'beforeUpdate', 'updated',
		'beforeDestroy', 'destroyed',
		'render', /* not a lifecyle hook, but handled identically */
]

export function lifecycle(target: any, key: string): void {
	if (lifeCycleHooks.indexOf(key) === -1) {
		return warn(
			'The @lifecycle decorator only ' +
			'applies to valid lifecycle hooks ' +
			'and the render method.',
		)
	}

	addDecorator(key, options => {
		(options as any)[key] = Object.getOwnPropertyDescriptor(target, key).value

		return true
	})
}

export const Lifecycle = lifecycle
