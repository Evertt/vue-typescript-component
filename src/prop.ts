import Vue = require('vue')
import { addDecorator } from './decorators'

export function prop(target: any, key: string): void {
	addDecorator(key, options => {
		let property: Vue.PropOptions = {}

		if (typeof target[key] === 'object') {
			property.default = () => target[key]
		} else {
			property.default = target[key]
		}

		if (property.type === undefined) {
			try {
				property.type = Object.getPrototypeOf(target[key]).constructor
			} catch (error) {
				// property.type = target[key].__proto__ ? target[key].__proto__.constructor : target[key].constructor
			}
		}

		if (property.default === undefined) {
			property.required = true
		}

		(options.props || (options.props = {}) as any)[key] = property
	})
}

export const Prop = prop
