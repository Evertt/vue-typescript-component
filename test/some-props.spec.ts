/// <reference path='../node_modules/@types/jest/index.d.ts' />

import Vue = require('vue')
import * as vts from '../src/vue-typescript-component'

class Foo {
	someString = 'abc'
	someNumber = 123
	someArray = [123, 'abc']
	someObject = {aString: 'abc', aNumber: 123, aArray: [123, 'abc']}
}

function getabc() {
	return 'abc'
}

@vts.component
class JustProps extends Vue {
	@vts.prop someString = 'abc'
	@vts.prop someNumber = 123
	@vts.prop someArray = [123, 'abc']
	@vts.prop someObject = {aString: 'abc', aNumber: 123, aArray: [123, 'abc']}
	@vts.prop someFoo = new Foo()
	@vts.prop someFunction = getabc

	render(createElement: typeof Vue.prototype.$createElement) {
		return createElement('div', JSON.stringify({
			someString: this.someString,
			someNumber: this.someNumber,
			someArray: this.someArray,
			someObject: this.someObject,
			someFoo: this.someFoo,
			someFunction: typeof(this.someFunction) === 'function' ? this.someFunction() : 'NOT A FUNCTION',
		}, null, '  '))
	}
}

it('creates the expected options', () => {
	expect((<any>JustProps).vueComponentOptions).toMatchSnapshot()
})

it('accepts the intended props', () => new Promise((resolve, reject) => {
	const vm = new (Vue.extend((<any>JustProps).vueComponentOptions))({
		propsData: {
			someString: 'abc',
			someNumber: 123,
			someArray: [123, 'abc'],
			someObject: {aString: 'abc', aNumber: 123, aArray: [123, 'abc']},
			someFoo: new Foo(),
			someFunction: getabc,
		},
	}).$mount(document.createElement('div'))
	vm.$nextTick(() => {
		expect(vm.$el.innerHTML).toMatchSnapshot()
		resolve()
	})
}))

it('rejects props with wrong types', () => {
	const mockConsoleError = jest.fn()
	console.error = mockConsoleError
	new (Vue.extend((<any>JustProps).vueComponentOptions))({
		propsData: {
			someString: 123,
			someNumber: 'abc',
			someArray: {aString: 'abc', aNumber: 123, aArray: [123, 'abc']},
			someObject: [123, 'abc'],
			someFoo: getabc,
			someFunction: new Foo(),
		},
	}).$mount(document.createElement('div'))
	expect(mockConsoleError.mock.calls).toMatchSnapshot()
})
