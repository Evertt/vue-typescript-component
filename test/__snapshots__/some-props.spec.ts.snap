exports[`test accepts the intended props 1`] = `
"{
  \"someString\": \"abc\",
  \"someNumber\": 123,
  \"someArray\": [
    123,
    \"abc\"
  ],
  \"someObject\": {
    \"aString\": \"abc\",
    \"aNumber\": 123,
    \"aArray\": [
      123,
      \"abc\"
    ]
  },
  \"someFoo\": {
    \"someString\": \"abc\",
    \"someNumber\": 123,
    \"someArray\": [
      123,
      \"abc\"
    ],
    \"someObject\": {
      \"aString\": \"abc\",
      \"aNumber\": 123,
      \"aArray\": [
        123,
        \"abc\"
      ]
    }
  },
  \"someFunction\": \"abc\"
}"
`;

exports[`test creates the expected options 1`] = `
Object {
  "computed": Object {},
  "data": [Function],
  "methods": Object {},
  "name": "JustProps",
  "props": Object {
    "someArray": Object {
      "default": [Function],
      "type": [Function],
    },
    "someFoo": Object {
      "default": [Function],
      "type": [Function],
    },
    "someFunction": Object {
      "default": [Function],
      "type": [Function],
    },
    "someNumber": Object {
      "default": 123,
      "type": [Function],
    },
    "someObject": Object {
      "default": [Function],
      "type": [Function],
    },
    "someString": Object {
      "default": "abc",
      "type": [Function],
    },
  },
  "render": [Function],
  "watch": Object {},
}
`;

exports[`test rejects props with wrong types 1`] = `
Array [
  Array [
    "[Vue warn]: Invalid prop: type check failed for prop \"someString\". Expected String, got Number. 
(found in root instance)",
  ],
  Array [
    "[Vue warn]: Invalid prop: type check failed for prop \"someNumber\". Expected Number, got String. 
(found in root instance)",
  ],
  Array [
    "[Vue warn]: Invalid prop: type check failed for prop \"someArray\". Expected Array, got Object. 
(found in root instance)",
  ],
  Array [
    "[Vue warn]: Invalid prop: type check failed for prop \"someObject\". Expected Object, got Array. 
(found in root instance)",
  ],
  Array [
    "[Vue warn]: Invalid prop: type check failed for prop \"someFoo\". Expected Foo, got Function. 
(found in root instance)",
  ],
  Array [
    "[Vue warn]: Invalid prop: type check failed for prop \"someFunction\". Expected Function, got Object. 
(found in root instance)",
  ],
]
`;
